### PR TITLE
Collect usage about used js modules and outputs

### PR DESCRIPTION
**Pull request by codebien**
## What?

It collects usage stats about the js modules and outputs used in the test. It collects only the names of the used modules or outputs that are built-in in k6 binary.  

## Why?

It will help to refine our prioritization method based on more objective and data-driven decisions regarding a feature or a breaking change to introduce.

Updates https://github.com/grafana/k6/issues/2952

### CODE REVIEW

File: cmd/report.go
Comment (joanlopez) [inline cmd/report.go:38]: Would make sense to keep a list, even if manually, here to also get usage stats for the extensions that we do own? At least until we find a better way to do so.

From the product perspective, I guess it could make lot of sense to know what extensions worth some attention and which ones not.

File: cmd/report.go
Comment (joanlopez) [inline cmd/report.go:56]:
```suggestion
	publicOutputs := make([]string, 0, len(builtinOutputs))
```

Nit; I think `publicOutputs` could only be as big as `builtinOutputs`, despite of the length of `outputs`.

File: cmd/report.go
Comment (joanlopez) [inline cmd/report.go:68]: Is this accurate? I'm not expert enough yet, but looks like the value returned under the hood `initializedVUs` can be decreased, so I'm wondering whether this is accurate or not depending on the scenarios configured, and whether this represents a "max" value.

File: cmd/report.go
Comment (joanlopez) [inline cmd/report.go:95]: I know the use of the `DefaultClient` is generally discouraged, because it lacks the definition of timeouts and other proper settings for production environments. Is it safe to use it here?

File: cmd/report_test.go
Comment (joanlopez) [inline cmd/report_test.go:54]: Would make sense to switch to  (or add) something like?

```suggestion
	assert.Equal(t, consts.Version, r.Version)
```

File: cmd/report_test.go
Comment (joanlopez) [inline cmd/report_test.go:36]: Nit; If I'm not wrong, this is only needed to have `constant-vus` as executor. However, imho this is a bit confusing, because it's used nowhere, and differs from the actual execution (~10ms) and the actual amount of VUs (6).

So, I'd suggest to either make it match with the values below (perhaps extracting these values as a shared variable?), or just removing these lines and switching the executor type in the assertion, which is even simpler/shorter, thus easier to understand.

File: cmd/report.go
Comment (codebien) [reply cmd/report.go:38]:
```
xk6-client-tracing
xk6-notification
xk6-disruptor
xk6-sql
xk6-exec
xk6-kubernetes
xk6-ssh
xk6-loki
xk6-dashboard
xk6-output-influxdb
xk6-output-kafka
xk6-output-timescaledb
```
This is the list of extensions marked as official here: https://k6.io/docs/extensions/get-started/explore. None of them sounds so relevant at the moment to justify the risk of having this logic. Let me know if you have a different opinion.

Comment (joanlopez) [reply cmd/report.go:38]: Sounds fair! Thanks!

File: cmd/report.go
Comment (codebien) [reply cmd/report.go:68]: Yes, at least in terms of _planned_ max VUs, it should be accurate. The unique case for decreasing is in the externally controlled executor which is a special case.

As an alternative, we may use the latest generated `VUID` but it has the same issue of being an indirect logic, so if the VUID logic changes then the report would be impacted without any guarantee to not introduce a bug.

You can read more about the original decision here https://github.com/grafana/k6/pull/1343#discussion_r388141537

File: cmd/report.go
Comment (codebien) [reply cmd/report.go:95]: >it lacks the definition of timeouts
>Is it safe to use it here?

Specifically, regarding the timeout, we should be safe as we are passing the timeout-limited context.

In general, it would be a good potential improvement. However, not so far in the future, we have to migrate to the new Grafana's service API. Then it may be convenient to rewrite and refactor this client all at once.

File: cmd/report.go
Comment (joanlopez) [reply cmd/report.go:68]: Thanks so much, all clear now! :)

File: cmd/report.go
Comment (joanlopez) [reply cmd/report.go:95]: Fair enough, let's move on! Also, looks like this code already existed before, so 🚀

**Review by joanlopez — DISMISSED** [review]
I left a few comments, but none of them is blocking, so it looks good enough to be merged to me! 👍🏻 Thanks!

File: cmd/test_load.go
Comment (olegbespalov) [inline cmd/test_load.go:43]: Looking into the usage of this property in all cases, it sets the value using the value of the `runner,` which already exists as a `test.initRunner`. Am I missing something, or could we avoid introducing the duplication?

Comment (codebien) [reply cmd/test_load.go:43]: `test.initRunner` is a lib.Runner interface https://github.com/grafana/k6/blob/65b2f27d1a0612846467e48418f469984724e007/lib/runner.go#L54 and it doesn't provide direct access to the ModuleResolver type that is a specific of the [js.Runner](https://github.com/grafana/k6/blob/65b2f27d1a0612846467e48418f469984724e007/js/runner.go#L47-L48) implementation.

It sounds like this is not clear enough from the code, do you think a comment somewhere could help?

Comment (olegbespalov) [reply cmd/test_load.go:43]: Ah, I see 👍

No I prefer to avoid commenting on this. It's just a lack of the PR context. In Code it should be fine.

**Comment by codebien** [comment]
@mstoykov as you are working on refactoring some closer areas, it would be good to get a review from you as well.

File: cmd/report.go
Comment (olegbespalov) [inline cmd/report.go:51]: Nit: These output keys (cloud, csv, etc.) come from the map https://github.com/grafana/k6/blob/master/cmd/outputs.go#L26-L47. It's better to extract this as the variable/method and use it here. That way, we protect the future maintainers/contributors from forgetting to update this place. 

Some alternatives (e.g., using `const`) could also be considered, but the main idea is to couple the places through the code, not the string values.

**Review by olegbespalov — DISMISSED** [review]
LGTM, with a minor non-blocking nit.

**Review by mstoykov — APPROVED** [review]
LGTM! I did check commit by commit and did some :+1: on some comments that I agreed more strongly with :sweat_smile: 

My only concern around pushing every imported module has been adressed so I am happy.

I guess (as future improvement) we can also send whether and how many `k6/x/` modules were imported? Maybe even how many source modules were imported :shrug: . Just as a number.

Also, I guess at some point we might want to have a "whitelist" of `k6/x/' modules that are still reported, but that definitely should be done later and with more discussion.

File: cmd/report.go
Comment (codebien) [reply cmd/report.go]: Added a new proposal for addressing this in the last commits

**Review by mstoykov — APPROVED** [review]

File: cmd/outputs.go
Comment (codebien) [reply cmd/outputs.go:27]: >enums decrease code reading/maintaining abilities

The reason is to reduce the probability of introducing bugs in the future by changing the list of builtin outputs (adding or removing them). In this way, we can rely on a single source of truth across the code more than directly on strings.

>and it seems like the builtin prefix could be omitted

Do you mean only from the items' names or also from the type? Because having an `output` type when we already have the `output` package and `output.Output` type could create confusion if there isn't an element to differentiate them.

**Review by joanlopez — APPROVED** [review]

File: cmd/outputs.go
Comment (joanlopez) [reply cmd/outputs.go:27]: Honestly, as someone who has never used `enumer` before, I have to admit that the generated code looks a bit overwhelming. But as suggested before, I'm _generally_ in favor or reducing chances to introduce errors by reducing duplicity, and as `enumer` seems to be already used in the project, I'd say it looks good enough to be merged to me.

**Review by olegbespalov — APPROVED** [review]

**Pull request merged**
